File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        output = dummy
        carry = 0
        while l1 or l2 or carry:
            curr = 0
            if l1:
                curr += l1.val
                l1 = l1.next
            if l2:
                curr += l2.val
                l2 = l2.next
            if carry:
                curr += carry
                carry = 0
            if curr >= 10:
                curr -= 10
                carry = 1
            output.next = ListNode(curr)
            output = output.next
        return dummy.next
        
            


        

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode();
        ListNode curr = head;
        dummy.next = head;
        ListNode prevGroupPointer = dummy;
        int check = 1;
        while (curr != null) {
            // Found 1 group of length k
            ListNode firstNode = curr;
            ListNode nextPointer = curr;
            for (int i = 0; i < k; i++) {
                if (curr == null) {
                    check = 0;
                    break;
                }
                check = 1;
                curr = curr.next;
            }
            // This would reverse k nodes and set the prevGroupPointer to the last node
            // Last node is curr
            if (check == 0) {
                prevGroupPointer.next = firstNode;
                break;
            }
            ListNode next = null;
            ListNode prev = null;
            while (firstNode != curr) {
                next = firstNode.next;
                firstNode.next = prev;
                prev = firstNode;
                firstNode = next;
            }
            prevGroupPointer.next = prev;
            prevGroupPointer = nextPointer;
        }
        return dummy.next;
    }
}
